File: my-folder/0076-minimum-window-substring/solution.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        int l1=s.length(),l2=t.length(),len=INT_MAX;
        if(l1==0||l2==0)
            return "";
        map<char,int>maps;
        map<char,int>mapt;
        
        for(int i=0;i<l2;i++)
            mapt[t[i]]++;
        int l=0,r=0,count=0;
        string s1="";
        
        while(r<l1)
        {
            if(mapt.find(s[r])!=mapt.end())
            {
                maps[s[r]]++;
                if(maps[s[r]]<=mapt[s[r]])
                    count++;
            }
            
            if(count>=l2)
            {
                while(mapt.find(s[l])==mapt.end()||maps[s[l]]>mapt[s[l]])
                {
                    
                    maps[s[l]]--;
                    l++;
                }
                if(r-l+1<len)
                {
                    len=r-l+1;
                    s1=s.substr(l,len);
                }
            }
            r++;
        }
        return s1;
        
    }
};

File: my-folder/0076-minimum-window-substring/solution.java
class Solution {
    public String minWindow(String s, String t) {

        if (s.length() < t.length()) {
            return "";
        }
        HashMap<Character, Integer> map = new HashMap<>();

        for (int k = 0; k < t.length(); k++) {
            map.put(t.charAt(k), map.getOrDefault(t.charAt(k), 0) + 1);
        }

        int min = Integer.MAX_VALUE;
        String minS = "-1";

        int count = 0;
        count = map.size();

        int i = 0, j = 0;
        while (j < s.length()) {
            char c =s.charAt(j);
            if (map.containsKey(c)) {
                map.put(c, map.get(c) - 1);
                if (map.get(c) == 0)
                    count--;
            }
            if (count > 0) {
                j++;
            } else {
                while (count == 0) {
                    char charLeft = s.charAt(i);
                    if (min > j - i + 1) {
                        minS = s.substring(i, j + 1);
                        min =  j - i + 1;
                    }
                    
                    if (map.containsKey(charLeft)) {
                        map.put(charLeft, map.get(charLeft) + 1);
                        if (map.get(charLeft) == 1)
                            count++;
                    }
                    i++;
                }
                j++;
            }

        }
        if (minS.equals("-1"))
            return minS = "";
        return minS;

    }
}
